io/runes/runes.go

```diff
@@ -0,0 +1,29 @@
+package runes
+
+import "io"
+
+// Reader reads Unicode encoded bytes into data.
```

Comment:
```suggestion
// Reader reads Unicode (possibly) encoded bytes into runes.
// It generalizes UTF decoders (or any byte stream that encodes unicode code points/runes).
```

Reply:
Yeah this doc is not good but `data` here means the argument. Why the "possibly"? because this way it looks like it accepts non-unicode encoded in the decoder stream...

Just thinking now that maybe this interface has nothing to do with decoding unicode bytes... Like, it should write runes in the `[]rune` data argument, but the implementation can work on a rune stream, so no decoding actually happens...
Like software `a` writes runes to stdout and software `b` reads runes from stdin, then a implementation of `runes.Reader` just copies data in the same way as `io.Reader`.
Maybe the right thing here is just copy (or refer) the `io.Reader` docs. What do you think?

Reply:
Ah I missed the "data" parameter reference because it is a little odd that the documentation of the overall interface is referencing a specific argument of a method of the interface, that would be easier to spot if this was the doc for the actual Read method, which then has an data argument.

I was trying to not be too vague..but indeed in the end the interface models anything that reads runes, it can be a decoder from some underlying text encoding (most usual), just reading raw unicode data from somewhere like you said, etc. So I think we could copy most of io.Reader docs (if not all) indeed.

Reply:
> Ah I missed the "data" parameter reference because it is a little odd that the documentation of the overall interface is referencing a specific argument of a method of the interface, that would be easier to spot if this was the doc for the actual Read method, which then has an data argument.

yes, referencing method argument in the interface docs is nonsense.

> So I think we could copy most of io.Reader docs (if not all) indeed.

Ok.

Documenting an interface is strange because we have to specify how implementations should behave but still a lot of stuff could go wrong and there's no way for the compiler to catch misuses... like _the implementation must return io.EOF when there's no data available anymore_. It's odd having to say that in prose instead of a kind of constraint that the compiler can verify but I know it's undecidable whether a function ever returns a value, but thinking that maybe a non-Turing-complete language can have good features like verifying if interface implementations are correct.

Reply:
Yeah I had a similar insight when I was reading A Philosophy for Software Design:

https://github.com/katcipis/sophia/blob/master/notes/books/philosophy-software-design.md#as-part-of-interface

It talks a lot about how documentation is an integral part of interfaces (a first class citizen), I liked the idea, I would prefer a formal specification but it is really hard to do that formally for some cases, specially when you have full protocols where the interaction of multiple methods can affect an outcome (and are required to get something done). Like after calling Close it is invalid to call Read in an io.ReadCloser, I never found a language that allowed for this kind of interaction to be described formally on the interface itself, in a way that can validate if a implementation does that at build time, so you end up relying on good docs/specs/tests.

Coincidentally my example was also io.Reader, I actually like the interface a lot because at least it is much better than what I was used in the past (file abstraction on Python for example, which is crappy and leaks details and barelly works), and so far I never was able to have a better idea to model general purpose byte stream reading (which is useful as a base building block).

Besides interaction of different method calls operating on the same state, even for single methods there can be a lot of tricky stuff, you mentioned EOF handling, and from a decision from Go devs it is actually even more complex:

```
It may return the (non-nil) error from the same call or return the error
(and n == 0) from a subsequent call. An instance of this general case is that
a Reader returning a non-zero number of bytes at the end of the input stream
may return either err == EOF or err == nil. The next Read should return 0, EOF.
```

I have no idea how I could formalize something like this on an interface =P. This example is specially tricky because it relates to IO, like returning a value when there is no more available data, available data being IO/dynamic... how to formally generalize that there is no more data to be read ? (read from where ?). Reading a stream of bytes from the network is like an infinite turing machine tape 😆 . I think what Im trying to say is tha the constrained language would also require very constrained IO models, or else you end up being turing complete again...maybe there is no way to do stream programming and not be turing complete, you would need to give up on solving that class of problems...but Im not sure.

The best we can come up on cases like this is document it well on the interface + maybe provide a set of tests that operate on the interface level to guarantee that the behavior is there, but it can be tricky to have general purpose tests because error handling situations usually depend on implementation details to inject failures. Im thinking on having a good set of tests that act as a formal spec + can be used to test N implementations, but it is not very easy to cover everything with this.